Fix BC break in Windows readlink

johnstevenson opened the pull request:
GetFinalPathNameByHandleW is given a file handle to a symbolic link, rather than one to the target, and therefore returns an incorrect path.

This happens with a volume mount point. Given a volume at `D:` mounted to a folder at `C:\mnt`:
```php
<?php
   # PHP-7.3 and down
    readlink('C:\\mnt')  // returns 'D:\'

   # PHP-7.4 and up
    readlink('C:\\mnt')  // returns 'C:\mnt'     
```
The readlink code was refactored in 91c905e83 with a backward-compatibility fix in 14628f1c5

I needed to add `ERROR_SYMLINK_NOT_SUPPORTED` to the `code_to_errno_map` in win32\winutil.c because debug builds were aborting.

I also added `ERROR_NOT_A_REPARSE_POINT` forthe same reason, which is set if DeviceIoControl is given a file handle that is not a reparse point (when FSCTL_GET_REPARSE_POINT is given as the dwIoControlCode).

It would appear that readlink on Windows will always succeed if given a file or directory path and return this path if it is not a link.

@weltling I would appreciate your input about this. Thanks.

krakjoe commented:
ping @cmb69 @weltling ... *a windows thing is happening*

cmb69 commented:
> It would appear that readlink on Windows will always succeed if given a file or directory path and return this path if it is not a link.

Which is pretty inconsistent with POSIX, where `readlink()` fails with `EINVAL` if the path is not a symlink. It seems to me that we should change this with PHP 8.

To avoid the BC break for PHP 7.4 this PR looks good to me; if `php_win32_ioutil_readlink_int()` fails, `CreateFileW()` without `FILE_FLAG_OPEN_REPARSE_POINT` should be appropriate.

weltling commented:
Looks good, just that i had no chance to test it yet. @johnstevenson could you add a test please, at least limited to AppVeyor?

readlink() indeed doesn't fail on regular files on Windows, that's the reason adding that BC part while the internal API tried to do the right thing. Not sure about 8.0, but 7.4 should be in any case compatible. For 8.0 should be discussed. Note also, that there are some other chunks like junctions, which have no dedicated API, but might be handled across APIs as if they were symlinks. Perhaps one would need more exact dedicated APIs for userspace to handle those cases.

@cmb69 if tests was ok on your side, please merge when it suits to you.

Thanks.

KalleZ commented on win32/ioutil.c:
```diff
@@ -1111,7 +1112,21 @@ PW32IO ssize_t php_win32_ioutil_readlink_w(const wchar_t *path, wchar_t *buf, si
 
 	if (ret < 0) {
-		/* BC */
+		/* BC - get a handle to the target (if path is a symbolic link) */
```
Please have declarations first (for code consistency)

johnstevenson replied:
Apologies, but I don't understand what you would like me to do here.

cmb69 replied:
Kalle suggests to move the newly inserted statements after the alreay existing declarations of `target`, `offset` and `target_len`. Having declarations before statements was mandatory for C89 compliance, and may still be preferable for current C standards (although that's likely to be controversial).

johnstevenson replied:
Okay, thanks got it. I'll fix this.

weltling replied:
C99 flags has been merged actually ;)

Thanks.

cmb69 commented:
The patch works for mount points created with `mountval`, but writing a portable test appears to be non-trivial (we'd likely have to parse the output of `mountval /?` to get a volume name, if there is even a second volume, etc.)

weltling commented:
That's why I meant AppVeyor, at least, but should be fine on desktop, too. De facto, English locale is assumed anyway, and also it would run as Administrator. That all can be checked in skip. Finding a non existent volume were not a big thing, too :) Fe see tests in ext/standard/tests/file/windows_acls. It definitely would make sense to ensure this kind of BC related changes.

Thanks.

johnstevenson commented:
I can write a simple test for 7.4, checking that a regular file, proper symlink (because we have admin priviledges on Appveyor, but would fail without this  [generally]), and a junction all succeed.

But should there be a deprecation notice, if we are going to change this in 8.0?

Also, we would need to have a plan because of inconsistencies. For example, `is_link` only returns true if given a symbolic link and fails on `IO_REPARSE_TAG_MOUNT_POINT`. So what should `readlink` do in the future?

cmb69 commented:
> I can write a simple test for 7.4, checking that a regular file, proper symlink (because we have admin priviledges on Appveyor, but would fail without this [generally]), and a junction all succeed.

It seems to me that these cases already succeed without this PR; only volume mount points would need this PR.

johnstevenson commented:
@cmb69 I was thinking more about if we are going to change this. And I don't know how to create a volume mount point easily for a test!

weltling commented:
@johnstevenson please lets focus on fixing the BC compatible way. Possible inconsistencies including those mentioned should be better handled in a collective patch and discussed first. Just changing the behavior without having a strategy is just another inconsistency :) And, probably it might be now easier handled in user space with ffi, anyway.

@johnstevenson to programmatically test

- `mountvol c:\ /L` to find out the volume
- create an empty directory,
- `mountvol dir volume`
- run the tests needed
- `mountvol dir /D`

Need administrator, so should be checked in skip section, should be fine for AppVeyor in any case.

Thanks.

johnstevenson commented:
@weltling Thanks (hadn't realized you could mount a volume to a folder within the same volume). I'll add this test.

For the future, where is the best place to discuss a strategy for this stuff?

johnstevenson commented:
I've added a compatibility test and fixed an additional BC break (where `readlink` on a relative symlink returned the relative path, rather than the absolute one).

weltling commented:
@johnstevenson merged as b37a6aafce721d289cfa004bfc8120796e01de60. And thanks for catching the other bug, probably should merge the test back into 7.2.

Regarding the discussion question - usually people use mailing lists. You can draft your ideas on the wiki, or on a gist at least. For the lists - internals-win or internals would be appropriate.

Thanks!
